core/chains/evm/logpoller/orm.go

```diff
@@ -386,25 +393,96 @@ type Exp struct {
 }
 
 func (o *DSORM) SelectUnmatchedLogIDs(ctx context.Context, limit int64) (ids []uint64, err error) {
-	query := `
-		SELECT l.id FROM evm.logs l LEFT JOIN (
+	batchLogsSubQuery := `SELECT id, evm_chain_id, address, event_sig FROM evm.logs
+                                WHERE evm_chain_id = $1 AND block_number >= $2 AND block_number <= $3`
+
+	query := fmt.Sprintf(`
+		SELECT l.id FROM (%s) l LEFT JOIN (
 			SELECT evm_chain_id, address, event
 			FROM evm.log_poller_filters
 				WHERE evm_chain_id = $1
 				GROUP BY evm_chain_id, address, event
 		) r ON l.evm_chain_id = r.evm_chain_id AND l.address = r.address AND l.event_sig = r.event
 		WHERE l.evm_chain_id = $1 AND r.evm_chain_id IS NULL
-	`
+	`, batchLogsSubQuery)
 
-	if limit == 0 {
-		err = o.ds.SelectContext(ctx, &ids, query, ubig.New(o.chainID))
+	latestBlock, err := o.SelectLatestBlock(ctx)
+	if err != nil {
 		return ids, err
 	}
-	err = o.ds.SelectContext(ctx, &ids, fmt.Sprintf("%s LIMIT %d", query, limit), ubig.New(o.chainID))
+
+	_, err = o.ExecPagedQuery(ctx, limit, latestBlock.FinalizedBlockNumber, func(lower, upper int64) (numResults int64, err2 error) {
+		var rowIDs []uint64
+		err2 = o.ds.SelectContext(ctx, &rowIDs, query, ubig.New(o.chainID), lower, upper)
+		// dedupe rowIDs before appending them to results
+		m := make(map[uint64]bool)
+		for _, id := range rowIDs {
+			if _, val := m[id]; !val {
+				m[id] = true
+				ids = append(ids, id)
```

Comment:
This looks a bit hacky, doesn't it? From the lambda function we are modyfing slice from the outer scope. Shouldn't we make `ExecPagedQuery` more configurable to just return it?

Reply:
I wonder if something like this would make it simpler in this case
```go
	_, resultIDs, err := o.ExecPagedQuery(ctx, limit, latestBlock.FinalizedBlockNumber, func(lower, upper int64) (numResults int64, acc uint64, err2 error) {
		var rowIDs []uint64
		err2 = o.ds.SelectContext(ctx, &rowIDs, query, ubig.New(o.chainID), lower, upper)
		return int64(len(rowIDs)), rowIDs, err2
	})
	
	uniqueIds := make(map[uint64]struct{}{}, len(resultIDs))
	for _, id := range resultIDs {
		uniqueIds[id] = struct{}{}
	}

	return maps.Keys(uniqueIds), err
```

```go
// acc is a simple accumulator, it's joined by the ExecPagedQuery function, meaning that 
// finalAcc := acc_1 + acc_2 + ... + acc_N 
//If generics are the problem it could be `acc []interface{}`
func ExecPagedQuery[T comparable](ctx context.Context, limit int64, end int64, query func(lower, upper int64) (int64, []T, error)) (numResults int64, acc []T, err error) {}
```

Then you should be able to use `ExecPagedQuery` with any query returning any results

Reply:
In which cases is it possible to get log ID duplicates?
Does `ExecPagedQuery` not go through distinct block ranges?

Reply:
@dhaidashenko There will be duplicates whenever a log matches more than one filter, due to the JOIN with the filters table.

Reply:
@dhaidashenko Actually, I guess you're right... they can't match with multiple filters in the Unmatched Logs query since they won't match with any filter, and the Excess Logs query has a GROUP BY id that should get rid of them. Must have gotten confused!

Reply:
> // acc is a simple accumulator, it's joined by the ExecPagedQuery function, meaning that // finalAcc := acc_1 + > acc_2 + ... + acc_N //If generics are the problem it could be `acc []interface{}` func ExecPagedQuery[T > comparable](ctx context.Context, limit int64, end int64, query func(lower, upper int64) (int64, []T, error)) > (numResults int64, acc []T, err error) {}

I thought of using generics for this, but it unfortunately doesn't allow them on methods. I'll see if I can get it to work with []interface{}

Reply:
I tried doing this with `[]interface{}` but it doesn't seem possible without some pretty heavy use of `reflect`, due to not being able to directly append to or cast a slice of an unknown type:
```
if reflect.TypeOf(rows).Kind() == reflect.Slice {
                    finalAcc = reflect.AppendSlice(reflect.ValueOf(finalAcc), reflect.ValueOf(rows))
                } else {
```
I decided that seems even more hacky than the closure. Maybe should have just gone back to my original at that point, but I ended up finding a way to do it with generics after all. It's also less than ideal, because I had to remove it from the `ORM` in order to use the generics which means you have to pass around the data source and chain id. Overall, I think it's an improvement though, because it seems less error prone and more flexible for adapting it to other purposes in the future.